### PR TITLE
Feat/client component

### PR DESCRIPTION
**Pull request by graykode**

### CODE REVIEW

File: contexts/currency-context.tsx [hunk 1]
```diff
@@ -34,11 +34,14 @@ export const CurrencyProvider = ({ children }: React.PropsWithChildren<{}>) => {
...
   const { address: userAddress } = useAccount()
   const { data: balance } = useBalance({ address: userAddress })
+  const { data: currencies } = useQuery(['currencies'], async () => {
+    return fetchCurrencies()
+  })
 
   const { data: prices } = useQuery(
     ['prices'],
     async () => {
-      const currencyAddresses = (await fetchCurrencies()).map(
```
Comment (onemonster) [inline contexts/currency-context.tsx:41]: Let's use `fetchCurrencies` here. It's better to have an extra api call if it means we can remove the dependency to the useQuery above. Code that causes side effects are bad, and code that depends on side effects are even worse. It's the path to spaghetti code.

File: model/market.ts [hunk 2]
```diff
@@ -233,9 +246,50 @@ export const calculateTotalDeposit = (
         initialDeposit,
       ))
     }
-    initialDeposit = markets.reduce((acc, market, i) => acc + amountOuts[i], 0n)
+    initialDeposit = amountOuts.reduce((a, b) => a + b, 0n)
```
Comment (onemonster) [inline model/market.ts:249]: `amountOuts.reduce((acc, val) => acc + val, 0n)`

File: model/market.ts [hunk 3]
```diff
@@ -233,9 +246,50 @@ export const calculateTotalDeposit = (
...
+  markets: Market[],
+  initialDeposit: bigint,
+  currentTimestamp: number,
+): {
+  epochStart: Date
+  epochEnd: Date
+  proceeds: bigint
+  apy: number
+} => {
+  if (
```
Comment (onemonster) [inline model/market.ts:267]: `some` looks more fitting in this situation.
```
    markets.some(
      (market) =>
        !isAddressEqual(
          market.quoteToken.address,
          substitute.address as `0x${string}`,
        ),
    )
```

File: pages/deposit/[symbol].tsx [hunk 4]
```diff
@@ -72,11 +57,89 @@ const Deposit: NextPage<InferGetStaticPropsType<typeof getStaticProps>> = ({
...
+        ? maxValue.toString()
+        : value,
+    )
+  }, [asset.underlying.address, asset.underlying.decimals, balances, value])
+
+  const setMaxValue = useCallback(() => {
+    setValue(
+      new BigNumber(balances[asset.underlying.address].toString())
+        .div(10 ** asset.underlying.decimals)
+        .toString(),
```
Comment (onemonster) [inline pages/deposit/[symbol].tsx:78]: `toString` might create weird results such as `1e27`.

```
    setValue(
      formatUnits(
        balances[asset.underlying.address] ?? 0n,
        asset.underlying.decimals,
      ),
    )
```

File: contexts/currency-context.tsx (hunk 1 above)
Comment (graykode) [reply contexts/currency-context.tsx]: Reverted: https://github.com/clober-dex/coupon.finance/pull/25/commits/c198f3e9f1019719480cc8cb842d483dd62dd34b

File: model/market.ts (hunk 2 above)
Comment (graykode) [reply model/market.ts]: Nice https://github.com/clober-dex/coupon.finance/pull/25/commits/7a6f1b9f574089137e3cd302f6a50b3080bb7b98

File: model/market.ts (hunk 3 above)
Comment (graykode) [reply model/market.ts:267]: https://github.com/clober-dex/coupon.finance/pull/25/commits/a22bdefe07066d04a9528243ec82943f136f5239

File: contexts/deposit-context.tsx [hunk 5]
```diff
@@ -119,7 +143,27 @@ export const DepositProvider = ({ children }: React.PropsWithChildren<{}>) => {
             },
           ],
         })
-        const { request } = await publicClient.simulateContract({
+        // const { request } = await publicClient.simulateContract({
```
Comment (onemonster) [inline contexts/deposit-context.tsx:146]: Revert back to using simulate after you're done debugging

File: contexts/permit-context.tsx [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+const Context = React.createContext<PermitContext>({
+  permit: () =>
+    Promise.resolve({
+      r: zeroBytes32 as `0x${string}`,
+      s: zeroBytes32 as `0x${string}`,
+      v: 0,
+    }),
+})
+
+export const PermitProvider = ({ children }: React.PropsWithChildren<{}>) => {
```
Comment (onemonster) [inline contexts/permit-context.tsx:34]: I think `permit` could just be a util function. Using context for this is overkill.

File: pages/deposit/[symbol].tsx [hunk 7]
```diff
@@ -116,10 +153,15 @@ const Deposit: NextPage<InferGetStaticPropsType<typeof getStaticProps>> = ({
...
                       value={value}
                       onValueChange={setValue}
+                      onBlur={onBlurValue}
                       placeholder="0.0000"
                     />
                     <div className="text-gray-400 dark:text-gray-500 text-xs sm:text-sm">
-                      ~$0.0000
+                      <ClientComponent>
+                        <>
+                          ~$ {Number(value) * prices[asset.underlying.address]}
```
Comment (onemonster) [inline pages/deposit/[symbol].tsx:162]: Let's use `formatDollarValue` in `utils/numbers.ts`

File: utils/epoch.ts [hunk 8]
```diff
@@ -0,0 +1,18 @@
+export const MAX_EPOCH = 4
```
Comment (onemonster) [inline utils/epoch.ts:1]: `MAX_EPOCHS`

File: contexts/deposit-context.tsx (hunk 5 above)
Comment (graykode) [reply contexts/deposit-context.tsx]: Done

File: utils/epoch.ts (hunk 8 above)
Comment (graykode) [reply utils/epoch.ts]: https://github.com/clober-dex/coupon.finance/pull/25/commits/54379398c2c07dc28f6b21595f2d6a61de22b2e1

**Pull request merged**